rpcs3/rpcs3qt/main_window.cpp

```diff
@@ -836,7 +836,58 @@ void main_window::InstallPup(QString file_path)
 	}
 }
 
-void main_window::HandlePupInstallation(QString file_path)
+void main_window::ExtractPup()
+{
+	const QString path_last_pup = m_gui_settings->GetValue(gui::fd_install_pup).toString();
+	QString file_path = QFileDialog::getOpenFileName(this, tr("Select PS3UPDAT.PUP To extract"), path_last_pup, tr("PS3 update file (PS3UPDAT.PUP);;All pup files (*.pup);;All files (*.*)"));
+
+	if (file_path.isEmpty())
+	{
+		return;
+	}
+
+	QString dir = QFileDialog::getExistingDirectory(this, tr("Extraction Directory"), QString{}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
+
+	if (!dir.isEmpty())
+	{
+		HandlePupInstallation(file_path, dir);
+	}
+}
+
+void main_window::ExtractTar()
+{
+	if (!m_gui_settings->GetBootConfirmation(this))
+	{
+		return;
+	}
+
+	Emu.SetForceBoot(true);
+	Emu.Stop();
+
+	const QString path_last_tar = m_gui_settings->GetValue(gui::fd_ext_tar).toString();
+	QStringList files = QFileDialog::getOpenFileNames(this, tr("Select TAR To extract"), path_last_tar, tr("All tar files (*.tar *.tar.aa.*);;All files (*.*)"));
+
+	if (files.isEmpty())
+	{
+		return;
+	}
+
+	QString dir = QFileDialog::getExistingDirectory(this, tr("Extraction Directory"), QString{}, QFileDialog::ShowDirsOnly | QFileDialog::DontResolveSymlinks);
+
+	if (dir.isEmpty())
+	{
+		return;
+	}
+
+	m_gui_settings->SetValue(gui::fd_ext_tar, QFileInfo(files[0]).path());
+
+	for (const QString& file : files)
+	{
+		extract_tar(sstr(file), sstr(dir) + '/');
+	}
```

Comment:
no it makes sense to add a progress dialog.
```suggestion
	progress_dialog pdlg(tr("TAR Extraction"), tr("Extracting encrypted TARs\nPlease wait..."), tr("Cancel"), 0, static_cast<int>(files.size()), false, this);
	pdlg.show();

	for (const QString& file : files)
	{
		if (pdlg.wasCanceled()) break;
		QApplication::processEvents();

		if (!extract_tar(sstr(file), sstr(dir) + '/'))
		{
			pdlg.SignalFailure();
			pdlg.setHidden(true);
			QMessageBox::critical(this, tr("Tar extraction failed"), tr("The following TAR file could not be extracted:\n%1").arg(file));
			break;
		}

		pdlg.SetValue(pdlg.value() + 1);
	}
```

Reply:
I modified it to be more lenient for solitary failures.

Reply:
I modified it to be more lenient for solitary failures.